### ISSUE TITLE
Image layouts specified in descriptor writes are not tracked

### ISSUE THREAD

**Issue by Ekzuzy** [KhronosGroup/Vulkan-LoaderAndValidationLayers#1454]
deferredshadows sample from Sascha Willems Vulkan at github has the following use case:

Render pass is set as follows:

initialLayout = VK_IMAGE_LAYOUT_UNDEFINED
finalLayout = VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL

but after the geometry is rendered using the above render pass, image is bound to a descriptor set, in which the following layout is used:

VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL

Despite image being in the wrong layout, validation layers don't report any errors.

*************

The same sample uses layered depth image (with 3 layers and a VK_IMAGE_VIEW_TYPE_2D_ARRAY view). Render pass, in which depth data is rendered into this image, has similar problems with layout transitions. But it can be fixed and we can use the following values:

initialLayout = VK_IMAGE_LAYOUT_UNDEFINED
finalLayout = VK_IMAGE_LAYOUT_DEPTH_STENCIL_READ_ONLY_OPTIMAL

with subpass layout set to:

VK_IMAGE_LAYOUT_DEPTH_STENCIL_ATTACHMENT_OPTIMAL

and the following layout specified in descriptor set (the same as final layout):

VK_IMAGE_LAYOUT_DEPTH_STENCIL_READ_ONLY_OPTIMAL

Unfortunately, with the above setup, validation layers report many problems:

DS(ERROR): object: 0x154b3500e58 type: 6 location: 4381 msgCode: 6: Cannot submit cmd buffer using image (0x6a) [sub-resource: aspectMask 0x4 array layer 0, mip level 0], with layout VK_IMAGE_LAYOUT_UNDEFINED when first use is VK_IMAGE_LAYOUT_DEPTH_STENCIL_ATTACHMENT_OPTIMAL.
DS(ERROR): object: 0x154b3500e58 type: 6 location: 4381 msgCode: 6: Cannot submit cmd buffer using image (0x6a) [sub-resource: aspectMask 0x4 array layer 1, mip level 0], with layout VK_IMAGE_LAYOUT_UNDEFINED when first use is VK_IMAGE_LAYOUT_DEPTH_STENCIL_ATTACHMENT_OPTIMAL.
DS(ERROR): object: 0x154b3500e58 type: 6 location: 4381 msgCode: 6: Cannot submit cmd buffer using image (0x6a) [sub-resource: aspectMask 0x4 array layer 2, mip level 0], with layout VK_IMAGE_LAYOUT_UNDEFINED when first use is VK_IMAGE_LAYOUT_DEPTH_STENCIL_ATTACHMENT_OPTIMAL.
DS(ERROR): object: 0x154b3500e58 type: 6 location: 4381 msgCode: 6: Cannot submit cmd buffer using image (0x6a) [sub-resource: aspectMask 0x4 array layer 0, mip level 0], with layout VK_IMAGE_LAYOUT_DEPTH_STENCIL_READ_ONLY_OPTIMAL when first use is VK_IMAGE_LAYOUT_DEPTH_STENCIL_ATTACHMENT_OPTIMAL.
DS(ERROR): object: 0x154b3500e58 type: 6 location: 4381 msgCode: 6: Cannot submit cmd buffer using image (0x6a) [sub-resource: aspectMask 0x4 array layer 1, mip level 0], with layout VK_IMAGE_LAYOUT_DEPTH_STENCIL_READ_ONLY_OPTIMAL when first use is VK_IMAGE_LAYOUT_DEPTH_STENCIL_ATTACHMENT_OPTIMAL.
DS(ERROR): object: 0x154b3500e58 type: 6 location: 4381 msgCode: 6: Cannot submit cmd buffer using image (0x6a) [sub-resource: aspectMask 0x4 array layer 2, mip level 0], with layout VK_IMAGE_LAYOUT_DEPTH_STENCIL_READ_ONLY_OPTIMAL when first use is VK_IMAGE_LAYOUT_DEPTH_STENCIL_ATTACHMENT_OPTIMAL.

It's strange, because similar setup for color or depth images with only 1 layer doesn't yield any errors. This suggests that layouts of layered images are incorrectly tracked in validation layers.

The above sample can be found here:

https://github.com/SaschaWillems/Vulkan

with affected files being:

https://github.com/SaschaWillems/Vulkan/blob/master/base/vulkanframebuffer.hpp
https://github.com/SaschaWillems/Vulkan/blob/master/deferredshadows/deferredshadows.cpp

**Comment by SaschaWillems** [comment]
[Discussion on this issue at my repo ](https://github.com/SaschaWillems/Vulkan/pull/238#issuecomment-277985991)

Changing the example as noted in the PR by @Ekzuzy makes it work as expected but then the validation messages noted here show up, while without the changes no validation errors are shown but the example won't render correct.

**Comment by tobine** [comment]
Alright, I'm able to run the deferredshadows demo on my Linux dev box with the latest validation layers. I went through the demo code and it appears that *almost* all of the changes from @Ekzuzy have been incorporated into the code I'm running. When I run on my dev setup I am **not** seeing any of the validation errors noted.
So, is there still suspected to be a validation issue with the second example given above? If so, what changes are required from the current latest demo code to make the code work and also trigger errors with latest validation code? I can believe we have an issue here, but need help in replicating.

Regarding the first issue noted, where descriptor set layout is not being validated, this is a hole. I filed #1476 as a separate issue to just track that missing case.

**Comment by tobine** [comment]
I take back my previous update about being unable to repro. I had hacked in explicit layers but was only enabling threading. Fixed my hack and now I'm seeing the validation errors. Will continue debug.

**Comment by tobine** [comment]
@Ekzuzy @SaschaWillems I believe there's an imageMemoryBarrier missing to transition the image in question from VK_IMAGE_LAYOUT_UNDEFINED to VK_IMAGE_LAYOUT_DEPTH_STENCIL_ATTACHMENT_OPTIMAL.

Here's what I'm seeing. The validation error in question is related to the depth attachment added to the frameBuffer in the deferredSetup() function.

Note that the code for that depth surface is very similar to the code for the depth surface in the shadowSetup(). However, the shadowSetup() code includes a call to vkTools::setImageLayout() that inserts an imageMemoryBarrier to transition that depth surface from its initial layout of VK_IMAGE_LAYOUT_UNDEFINED to the attachment.initialLayout that's it's assigned in addAttachment() which is VK_IMAGE_LAYOUT_DEPTH_STENCIL_ATTACHMENT_OPTIMAL;

I believe the depth attachment in deferredSetup() needs a similar transition.

Can you check this out and let me know what you think?
Thanks

**Comment by Ekzuzy** [comment]
Thanks for the info, but I'm not sure if validation error is related to the depth attachment specified in the deferredSetup() function. Validation layers report problem with a layered depth attachment, and such image is specified in the shadowSetup() function, not deferredSetup().

What's more. Render passes created in both functions define an UNDEFINED initial layout:

attachment.description.initialLayout = VK_IMAGE_LAYOUT_UNDEFINED;

Subpass layout is specified as (in case of depth attachments):

depthReference.layout = VK_IMAGE_LAYOUT_DEPTH_STENCIL_ATTACHMENT_OPTIMAL;

and the final layout is:

attachment.description.finalLayout = VK_IMAGE_LAYOUT_DEPTH_STENCIL_READ_ONLY_OPTIMAL;

With such setup all layout transitions are performed automatically by the driver, so the imageMemoryBarrier in the shadowSetup() function is, in my opinion, unnecessary.

**Comment by tobine** [comment]
@Ekzuzy yes you're right. That imageMemoryBarrier is not needed.
This is a corner-case bug in validation layer for depth/stencil images where the stencil aspect does not have its layout correct transitioned to the initialLayout for the RenderPass.

I have an initial fix as well as positive testcase to demonstrate the fix. I think I can improve the fix but I'll go ahead and submit the initial fix for review so we can get it on master and then I'll follow-up with some improvements.

I've also verified that the fix removes the error in the deferredShadows demo. Additionally I've verified that the "vkTools::setImageLayout()" call in shadowSetup() can be removed and the demo seems to run correctly with no validation errors. I'll leave it to you if you'd like to submit a PR for that change to Sascha's repo.
